### PR TITLE
MXCrypto: Interactive device verification

### PR DESCRIPTION
**Pull request by manuroe**
using SAS

https://github.com/vector-im/riot-ios/issues/2322

Missing things:
 - [x] OLMKit 3.1.0 pod release

Sequence diagram for the interactive device verification between alice & bob:

IMG_3735

### CODE REVIEW

File: MatrixSDK/Crypto/Verification/JSONModels/MXKeyVerificationStart.m [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+- (BOOL)isValid
+{
+    BOOL isValid = YES;
+
+    if (_method.length == 0
+        || _keyAgreementProtocols.count == 0
+        || _hashAlgorithms.count == 0
+        || [_hashAlgorithms containsObject:@"sha256"] == NO
+        || _messageAuthenticationCodes.count == 0
+        || [_messageAuthenticationCodes containsObject:kMXKeyVerificationSASMacSha256LongKdf] == NO
```
Comment (BillCarsonFr) [inline MatrixSDK/Crypto/Verification/JSONModels/MXKeyVerificationStart.m]: hkdf-hmac-sha256 is supposed to be mandatory, but for compatibilty reason request with only 'hmac-sha256' should be seen as valid

File: MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+                         fromUserId:event.sender
+                          andDevice:keyVerificationStart.fromDevice
+                               code:MXTransactionCancelCode.invalidMessage];
+        }
+
+        return;
+    }
+
+
+    MXDeviceVerificationTransaction *existingTransaction = [self transactionWithUser:keyVerificationStart.transactionId andDevice:keyVerificationStart.fromDevice];
```
Comment (BillCarsonFr) [inline MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m]: Param is withUser, but value passed is a transaction id

Comment (BillCarsonFr) [reply MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m]: Also this test is also done after the key download, maybe this one is not needed (unless for saving a download key)?

File: MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+            [existingTransaction cancelWithCancelCode:MXTransactionCancelCode.invalidMessage];
+            [self cancelTransaction:keyVerificationStart.transactionId
+                         fromUserId:event.sender
+                          andDevice:keyVerificationStart.fromDevice
+                               code:MXTransactionCancelCode.invalidMessage];
+            return;
+        }
+
+        
+        // We support only SAS at the moment
```
Comment (BillCarsonFr) [inline MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m]: Should test that method == "m.sas.v1" ?

Comment (manuroe) [reply MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m]: >  Oh it's done by the initWithOtherDevice?

yes. This is part of the validation process within `MXIncomingSASTransaction.initWithOtherDevice()`

File: MatrixSDK/Crypto/Verification/Transactions/SAS/MXSASTransaction.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#import "MXDeviceVerificationTransaction.h"
+
+#import "MXEmojiRepresentation.h"
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Constants
+
+FOUNDATION_EXPORT NSString * _Nonnull const kMXKeyVerificationMethodSAS;
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/Verification/Transactions/SAS/MXSASTransaction.h]: Same as above, do not use "k" prefix for public constants.

**Pull request merged**